### PR TITLE
Assessment groups improvements

### LINKED ISSUE #2115: Unable to leave group while a request is pending
After sending an invite for a group to the wrong person, a student is unable to cancel the formation of the group/leave the group.
This occurs at this screen:
IMG_2856

Potential fixes may be including a button to cancel the creation of the group altogether, or assigning someone else as the leader of the group.

### PR DESCRIPTION
**Pull request by damianhxy**
## Description

1. Add missing breadcrumbs for groups (`index`, `new`, `show`)
2. Replace links with path helpers where possible
3. Remove deadcode re: `@grouplessCUDs` which is never shown on `show` or `new` pages
4. For convenience, don't add instructor/CA to group when they create one (if self-assignment is disabled)
5. Groups clear their members upon destruction
6. Allow (only) instructors to boot students, i.e. even if they're already confirmed
7. Disband groups if the only members left are those that require group confirmation
8. UX improvements
  - Don't show "haven't confirmed group membership" message to instructors
  - Show names in addition to emails
  - Helper text to inform students when group is at max size or student assignment is not allowed
  - Flash when disbanding groups
  - Hide inapplicable UI elements (e.g. when student assignment is not allowed)
  - Improve explanation of behavior
9. Allow students to leave group while waiting for group approval (this is the one that resolves #2115)
10. Allow members to "add themselves" - essentially no effect for students, but useful for instructors who might want to add themselves to groups

## Motivation and Context

Closes #2115

## How Has This Been Tested?

Create assessment with group size 3 and student self-assignment enabled
- As `student1`, create a group with `student2`
- As `student3`, request to join the group
- (important) `student3` can choose to leave the group
- As `student1`, cancel invite to `student2` and then leave the group. Group will be disbanded.
- Verify that `student3` is now without a group
- As `student1`, create a new group with `student2`
- As `instructor`, kick `student1` - group not disbanded
- As `instructor`, kick `student2` - group disbanded

Create assessment with group size 3 and student self-assignment disabled
- As `student1`, see message that you can't create or join groups
- As `instructor`, create group with `student1`- observe that `student1` is automatically confirmed and `instructor` is not added
- As `instructor`, add yourself to the group successfully. Then kick yourself.
- As `student1` view group and see message that you can't add members to the group.

In general, test that behavior makes sense.

#### Instructor (in group), student-assignment allowed / not allowed
Note: when student-assignment allowed, need to self-approve. If not allowed, automatically approved.
Screenshot 2024-03-18 at 19 32 51
#### Instructor (not in group), student-assignment allowed / not allowed
Screenshot 2024-03-18 at 19 31 31
#### Confirmed student, student-assignment allowed
Screenshot 2024-03-18 at 19 29 26
#### Confirmed student, student-assignment not allowed
Screenshot 2024-03-18 at 19 29 33
#### Group-confirmed student, student-assignment allowed
Screenshot 2024-03-18 at 19 30 40
#### Group-confirmed student, student-assignment not allowed
Screenshot 2024-03-18 at 19 30 46
#### Member-confirmed student, student-assignment allowed
Screenshot 2024-03-18 at 19 31 08
#### Member-confirmed student, student-assignment not allowed
Screenshot 2024-03-18 at 19 31 15

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have run rubocop and erblint for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting
- [ ] My change requires a change to the documentation, which is located at [Autolab Docs](https://docs.autolabproject.com/)
- [ ] I have updated the documentation accordingly, included in this PR

### CODE REVIEW

File: app/views/groups/show.html.erb [hunk 1]
```diff
@@ -41,42 +46,47 @@
...
-        </li>
-      <% end %>
-    <% end %>
-  </ul>
-<% end %>
+<%# else: student assignment is allowed, but user is not a member of the group yet. We don't have to display anything %>
 
 <div style="margin-top: 1em">
-  <% if @group.is_member(@aud) then %>
+  <% if aud.group_id == id && (@assessment.allow_student_assign_group || @cud.instructor) %>
```
Comment (KesterTan) [inline app/views/groups/show.html.erb:89]: Getting an error here which does not exist on the master branch. I'm unable to create a new group or edit the group.

To replicate:
- Change the group limit of an existing assessment
- Create a group and invite one user
- Click create group

**Review by KesterTan — COMMENTED** [review]
Also self-assign groups is checked by default when creating a new assessment, only way to uncheck it is to do so in assessment settings. I think maybe it would be better to let instructors check/uncheck it before/while creating an assessment too?

File: app/views/groups/show.html.erb (hunk 1 above)
Comment (damianhxy) [reply app/views/groups/show.html.erb]: Fixed, mb

**Comment by damianhxy** [comment]
> Also self-assign groups is checked by default when creating a new assessment, only way to uncheck it is to do so in assessment settings. I think maybe it would be better to let instructors check/uncheck it before/while creating an assessment too?

Done:
Screenshot 2024-03-28 at 03 21 28

I also took the liberty of fixing the breadcrumbs for the creation page
Screenshot 2024-03-28 at 03 21 38

File: app/controllers/groups_controller.rb [hunk 2]
```diff
@@ -286,22 +288,47 @@ def leave
...
-      leaver.group = nil
-      leaver.membership_status = AssessmentUserDatum::UNCONFIRMED
       ActiveRecord::Base.transaction do
-        leaver.save!
-        @group.destroy! if @group.assessment_user_data.empty?
+        leaver.leave_group
+        if @group.is_effectively_empty
+          @group.destroy!
+          flash[:success] =
+            "Successfully left and disbanded group as all remaining members (if any) "\
```
Comment (KesterTan) [inline app/controllers/groups_controller.rb:325]: I don't think this works, when I create a group, invite someone who has not accepted and leave the group. The group is not disbanded, it still exists.

Comment (damianhxy) [reply app/controllers/groups_controller.rb:325]: This works as expected. The person you invited can still accept your invite and join the group.

File: app/controllers/groups_controller.rb [hunk 3]
```diff
@@ -250,19 +250,21 @@ def join
...
     if newMemberAUD.membership_status == AssessmentUserDatum::UNCONFIRMED ||
        newMemberAUD.group_id == @group.id
 
       newMemberAUD.group = @group
       newMemberAUD.membership_status |= AssessmentUserDatum::MEMBER_CONFIRMED
       newMemberAUD.save!
-      flash[:success] = "Membership confirmed!"
+      flash[:success] = "Member confirmed!"
+    else
+      flash[:error] = "You are already in a group."
```
[the remark this answers is not in the document]
Comment (damianhxy) [reply app/controllers/groups_controller.rb:267]: I think the current wording is fine - the issue is that the student is already in another group, regardless of membership status (CONFIRMED / GROUP_CONFIRMED / MEMBER_CONFIRMED). 

i.e. this error message also shows when they're a full member of a group, and then they try to join another group. So "already requested to join a group" isn't fully accurate.

**Review by KesterTan — APPROVED** [review]
Everything works as expected. LGTM

**Pull request merged**